src/common/pf_ppm.c

```diff
@@ -1023,23 +1025,60 @@ uint16_t pf_ppm_calculate_cyclecounter (
-   uint32_t transmission_interval_in_timebases;
    uint32_t number_of_transmissions;
+   const uint32_t transmission_interval_in_timebases = send_clock_factor * reduction_ratio;
 
    raw_cyclecounter = (32UL * timestamp) / 1000UL;
-   transmission_interval_in_timebases = send_clock_factor * reduction_ratio;
+
+   /* Integer division for truncation */
    number_of_transmissions =
-      raw_cyclecounter / transmission_interval_in_timebases; /* Integer division
-                                                                for truncation
-                                                              */
+      raw_cyclecounter / transmission_interval_in_timebases;
 
+   /* Now a multiple of transmission intervals */
    return number_of_transmissions *
-          transmission_interval_in_timebases; /* Now a
-                                                 multiple
-                                                 of
-                                                 transmission
-                                                 intervals
-                                               */
+          transmission_interval_in_timebases;
+}
+
+/**
+ * Calculate the next cyclecounter value from the previous one.
+ *
+ * The timebase is 1/32 of a millisecond (31.25 us).
+ *
+ * One cycle length is send_clock_factor x 31.25 us.
+ * The cyclenumber value is keeping track of which cycle we are in now.
+ *
+ * The reduction_ratio describes if transmit and receive should
+ * be done every cycle. A value of 1 corresponds to every cycle,
+ * while a value of 4 corresponds to transmit and receive every fourth cycle.
+ *
+ * For example, send_clock_factor=2 and reduction_ratio=4 gives a
+ * cycle time of 62.5 us, and a frame is sent every 250 us.
+ * In the first frame the cyclecounter value should be 0, and in the next should
+ * the value be 8.
+ *
+ * @param previous_cyclecounter In:   Previous cyclecounter value
+ * @param send_clock_factor     In:   Defines the length of a cycle, in units
+ *                                    of 31.25 us. Allowed 1 .. 128.
+ *                                    Typically powers of two.
+ * @param reduction_ratio       In:   Reduction ratio. If transmission should be
+ *                                    done every cycle. Allowed 1 .. 512.
+ *                                    Typically powers of two.
+ * @return Next cyclecounter value
+ */
+uint16_t pf_ppm_calculate_next_cyclecounter (
+   uint16_t previous_cyclecounter,
+   uint16_t send_clock_factor,
+   uint16_t reduction_ratio)
+{
+   uint16_t number_of_previous_transmissions;
+   const uint32_t transmission_interval_in_timebases = send_clock_factor * reduction_ratio;
+
+   /* Integer division for truncation */
+   number_of_previous_transmissions =
+      previous_cyclecounter / transmission_interval_in_timebases;
+
+   /* Now a multiple of transmission intervals */
+   return (number_of_previous_transmissions + 1) * transmission_interval_in_timebases;
```

Comment:
If I understand correctly following implementation is easy to follow and gives the same result :

return previous_cyclecounter + send_clock_factor * reduction_ratio;

In my opinion, introducing the transmission_interval_in_timebases intermediate variable just adds complexity.

Maybe I am missing something?

Reply:
This implementation makes sure that we always return a multiple of send_clock_factor * reduction_ratio, regardless of the previous_cyclecounter. See the testcases in this patch.
I will update the documentation.